### PR TITLE
interfaces/desktop: allow snaps to provide a desktop slot, granting access to system fonts and xdg-desktop-portal

### PR DESCRIPTION
**Pull request by jhenstridge**
This is a resubmission of #10609, with some additions.

At present, the only slot implementation for the desktop interface is an implicit slot on classic systems. In order to use GUI snaps on Ubuntu Core Desktop, we need something else to provide the slot. At the moment, our plan has been to have the `ubuntu-desktop-session` snap provide the slot, as it is already running many of the services this interface grants access to.

This PR is not a complete implementation of that idea, but starts by allowing application snaps to provide the slot and to provide access to host system fonts in this form.

As with the implicit slot on classic systems, this is bind mounting in font directories from the host system mount namespace. On a standard Ubuntu Core system, this will bind mount a number of empty directories into the plug-side snap's mount namespace as it doesn't ship any fonts.  On the Ubuntu Core Desktop prototype, we do have a selection of fonts available in the boot base snap though.

Ideally we'd instead be mounting from the fonts we know are visible to the slot-side snap. In practice, that would mean mounting from the slot snap's base snap.  This seems like it'd be a bit fragile though, since the AppArmor and mount rules for the plug snap won't be regenerated when the slot-side's base snap is upgraded.

### CODE REVIEW

**Comment by mardy** [comment]
Looks good to me so far. :-)

Just to make sure I understood the plan: on core systems, we will have some snaps (tha base one, and maybe also others in the future?) which provide fonts. When such snaps are installed, bind mounts will be created for the individual font files to make them appear under the system's `/user/share/fonts`. This directory is then recursively bind mounted into every snap using the desktop interface, so that the fonts will be accessible there too.

If the above is correct, then I don't understand you last remark about AppArmor: the apparmor rules for the plug snap will only need to grant access to `/usr/share/fonts/`, so why will they need to be regenerated?

**Comment by jhenstridge** [comment]
The scripts used to build the current Ubuntu Core Desktop prototype image can be found here:

https://github.com/canonical/ubuntu-core-desktop/

It differs from a regular Ubuntu Core device in that the boot file system comes from a `core20-gdm` base snap rather than `core20`. This base snap includes additional packages to enable the GDM login screen and the gnome-initial-setup user creation workflow. That effectively comprises a basic GNOME system.

The user session runs as a command exposed by the `ubuntu-desktop-session` snap that is included by the model definition. That snap doesn't actually contain any binary code or fonts: instead it uses `core20-gdm` as its base, and runs the binaries found there.

In it's current form, this PR works for the prototype image, because the base snap of ubuntu-desktop-session is the same as the boot base snap of the system as a whole. Ideally the set of fonts should depend on whatever snap provides the slot. But as I outlined this is a bit fragile if we wanted to use fonts from the "slot snap's base snap".

File: interfaces/builtin/desktop.go
Comment (pedronis) [inline interfaces/builtin/desktop.go:378]: same

File: interfaces/builtin/desktop.go
Comment (pedronis) [inline interfaces/builtin/desktop.go:41]: we might want to keep this as is and allow single snaps one by one to have this slot?

**Review by pedronis — COMMENTED** [review]
this looks reasonable, but if I understand correctly kind of assumes that the boot base and the base of the session snap should usually be the same? might there should be a comment about that? also I am a bit unsure if we really want any random snap to sport this slot, or we want to control which snap can sport it

File: interfaces/builtin/desktop.go
Comment (jhenstridge) [reply interfaces/builtin/desktop.go:41]: If the base declaration only allows `slot-snap-type: [core]`, then it is impossible to install an unpublished snap with the slot. It doesn't hit the "default allow" path for this case  like other declaration constraints. It'd be a pain to only be able to test a snap after it has been published and gained a snap-declaration from the store.

File: interfaces/builtin/desktop.go
Comment (mardy) [inline interfaces/builtin/desktop.go:117]: Unless my eyes are betraying me, the last two rules are identical to the first two.

File: interfaces/builtin/desktop.go
Comment (mardy) [inline interfaces/builtin/desktop.go:321]: This seems to be unused, I guess it's a leftover from another branch?

File: interfaces/builtin/desktop.go
Comment (alexmurray) [reply interfaces/builtin/desktop.go:117]: Your eyes are betraying you - the first 2 are Desktop the second two are Documents

File: interfaces/builtin/desktop.go
Comment (jhenstridge) [reply interfaces/builtin/desktop.go:321]: I've added an extra revision that sorts this out. I'd been testing these additional AppArmor rules ad-hoc, and noticed I'd missed this when rebuilding a new image.

**Comment by jhenstridge** [comment]
I've updated this PR adding the changes we needed to get xdg-desktop-portal working on core. The architecture looks something like this:

```mermaid
sequenceDiagram
  participant A as app
  participant U as unconfined
  participant S as desktop-session
  A->>U: call o.fd.portal.* API
  U->>S: call o.fd.impl.portal.* backend to show UI
  S->>U: response from UI
  U->>A: respond to app
```

In this case, all the UI and desktop integration provided by gnome-shell, xdg-desktop-portal-gtk, and xdg-desktop-portal-gnome runs confined in the context of the desktop-session snap, with unconfined xdg-desktop-portal and xdg-document-portal acting as an intermediary to the app.

File: interfaces/builtin/desktop.go
Comment (mardy) [reply interfaces/builtin/desktop.go:117]: Ah, I see, indeed! :-)

My last question then: if I understand correctly, these four rules are essentially the same as the two above having `label=unconfined`, and are needed to handle the case where the services are not running and need to be activated by their service name. I would make this more obvious by adding a sentence at the beginning of the comment block above, saying something like: "Replicate the rules above specifying the well-known service name this time; this is needed for those cases when the services need to be DBus-activated."

**Review by mardy — APPROVED** [review]
LGTM, a nitpick about a comment which could be more explicit.

File: interfaces/builtin/desktop.go
Comment (pedronis) [inline interfaces/builtin/desktop.go:46]: we can't express this atm but we probably would like to stop auto-connection if the slot is from an app. OTOH see the other point.

File: interfaces/builtin/desktop.go
Comment (pedronis) [reply interfaces/builtin/desktop.go:41]: we have hit similar issues before and we have other use cases. Basically there are slot for which we want to be careful about the apps that have them but OTOH we don't want to stop development of such snaps. The simplest if awkward approach is to change the checks we do for locally installed snaps to consider only the allow-installation rules. Then the strange:

```
allow-installation:
   slot-snap-type:
   - core
   - app
 deny-installation:
   slot-snap-type:
      - app
```

Would achieve what we intend.

A different approach would be to introduce something like "dangerous-app" that is a subset of app but it is awkward in different ways.

**Review by pedronis — COMMENTED** [review]
some more considerations about permissions

File: interfaces/builtin/desktop.go
Comment (jhenstridge) [reply interfaces/builtin/desktop.go:127]: For what it is worth, these rules along with their comments are just being moved within the file (to separate core+classic and classic-only cases), rather than being new rules. I'd prefer not to go through rewording things that are essentially unmodified.

**Review by robert-ancell — APPROVED** [review]
LGTM!

File: interfaces/builtin/desktop.go
Comment (jhenstridge) [reply interfaces/builtin/desktop.go:48]: The allow-installation/deny-installation pair seems to work with current snapd. I've also added policy unit tests to verify, and the spread test using an unpublished snap with a desktop slot is working.

File: interfaces/builtin/desktop.go
Comment (jhenstridge) [reply interfaces/builtin/desktop.go:56]: It would be appropriate to switch to using a `slot-snap-type` constraint here once #13062 lands. Given that we've got the installation constraint working, would it be best to leave this out entirely for now, or wait for the other PR to land?

**Review by pedronis — COMMENTED** [review]
this looks reasonable

File: interfaces/builtin/desktop.go
Comment (pedronis) [reply interfaces/builtin/desktop.go:56]: in #13062 we don't have an installation constraint, this is probably fine here at least for now

**Review by alexmurray — APPROVED** [review]
LGTM - thanks @jhenstridge

File: interfaces/builtin/desktop_test.go
Comment (pedronis) [inline interfaces/builtin/desktop_test.go:164]: do we need to keep some minimal testing for the coreSlot case here too?

Comment (jhenstridge) [reply interfaces/builtin/desktop_test.go:164]: These checks have been moved down to lines 195-197. As I removed the early exit for `!release.OnClassic`, it seemed easier to put all the coreSlot tests together.

**Comment by pedronis** [comment]
@jhenstridge this also needs a test re-run

**Comment by jhenstridge** [comment]
I've merged master to kick off another test run. I don't think any of the failures look related to my changes.

**Pull request merged**
